### PR TITLE
[Feature] User variable support use subquery expression

### PR DESCRIPTION
**Pull request by HangyuanLiu**
## What type of PR is this：
- [ ] bug
- [x] feature
- [ ] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：

Fixes #

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] I have added user document for my new feature or new function

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/qe/SetExecutor.java [hunk 1]
```diff
@@ -86,4 +148,21 @@ public void setSessionVars() throws DdlException {
             }
         }
     }
+
+    private static List<TVariableData> deserializerStatisticData(List<TResultBatch> sqlResult) throws TException {
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/qe/SetExecutor.java:152]:
```suggestion
    private static List<TVariableData> deserializerStatisticData(List<TResultBatch> sqlResult) throws TException {
```
????what the name?

File: fe/fe-core/src/main/java/com/starrocks/qe/SetExecutor.java [hunk 2]
```diff
@@ -54,7 +75,48 @@ private void setVariablesOfAllType(SetVar var) throws DdlException {
             // do nothing
             return;
         } else {
-            ctx.modifySessionVariable(var, false);
+            if (var.getType().equals(SetType.USER)) {
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/qe/SetExecutor.java:78]: extract to a method

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/ExpressionAnalyzer.java [hunk 3]
```diff
@@ -836,6 +839,47 @@ public Void visitInformationFunction(InformationFunction node, Scope context) {
         @Override
         public Void visitVariableExpr(VariableExpr node, Scope context) {
             try {
+                if (node.getSetType().equals(SetType.USER)) {
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/analyzer/ExpressionAnalyzer.java:842]: add comment about how to resolve type

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/ExpressionAnalyzer.java [hunk 4]
```diff
@@ -836,6 +839,47 @@ public Void visitInformationFunction(InformationFunction node, Scope context) {
...
+                    String variableValue = userVariable.getResolvedExpression().getStringValue();
+
+                    if (userVariable.getResolvedExpression() instanceof NullLiteral) {
+                        node.setType(variableType);
+                        node.setIsNull();
+                        return null;
+                    }
+
+                    node.setType(variableType);
+                    switch (variableType.getPrimitiveType()) {
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/analyzer/ExpressionAnalyzer.java:860]: can be support select bitmap/hll/array/json?

File: gensrc/thrift/Data.thrift [hunk 5]
```diff
@@ -104,3 +104,8 @@ struct TStatisticData {
     13: optional i64 meta_version
 }
 
+// Result data for user variable
+struct TVariableData {
+    1: optional bool isNull
+    2: optional string result
```
Comment (kangkaisen) [inline gensrc/thrift/Data.thrift:110]: binary  better than string. from our global dict thrift experience.

File: fe/fe-core/src/main/java/com/starrocks/analysis/Expr.java [hunk 6]
```diff
@@ -1334,11 +1337,15 @@ public final boolean isMonotonic() {
...
-        return ScalarOperatorToExpr.buildExprIgnoreSlot(scalarOperator,
-                new ScalarOperatorToExpr.FormatterContext(Maps.newHashMap()));
+        try {
+            ScalarOperator scalarOperator = SqlToScalarOperatorTranslator.translate(expr);
+            ScalarOperatorRewriter scalarRewriter = new ScalarOperatorRewriter();
+            // Add cast and constant fold
+            scalarOperator = scalarRewriter.rewrite(scalarOperator, ScalarOperatorRewriter.DEFAULT_REWRITE_RULES);
+            return ScalarOperatorToExpr.buildExprIgnoreSlot(scalarOperator,
+                    new ScalarOperatorToExpr.FormatterContext(Maps.newHashMap()));
+        } catch (UnsupportedException e) {
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/analysis/Expr.java:1347]: Why we couldn't throw exception directly here?

File: be/src/runtime/variable_result_writer.cpp [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+    }
+
+    Status status = _sinker->add_batch(result);
+
+    if (status.ok()) {
+        _written_rows += num_rows;
+        return status;
+    }
+
+    LOG(WARNING) << "Append user variable result to sink failed.";
```
Comment (kangkaisen) [inline be/src/runtime/variable_result_writer.cpp:81]: could also log the status to warning log

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/ExpressionAnalyzer.java (hunk 3 above)
Comment (HangyuanLiu) [reply fe/fe-core/src/main/java/com/starrocks/sql/analyzer/ExpressionAnalyzer.java:842]: done

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/ExpressionAnalyzer.java (hunk 4 above)
Comment (HangyuanLiu) [reply fe/fe-core/src/main/java/com/starrocks/sql/analyzer/ExpressionAnalyzer.java:861]: support json, bitmap/hll/array not support

File: gensrc/thrift/Data.thrift (hunk 5 above)
Comment (HangyuanLiu) [reply gensrc/thrift/Data.thrift]: done

File: be/src/runtime/variable_result_writer.cpp (hunk 7 above)
Comment (HangyuanLiu) [reply be/src/runtime/variable_result_writer.cpp]: done

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 107 / 189 (56.61%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/analysis/OutFileClause.java|0|1|00.00%|[227]|
|:large_blue_circle:|com/starrocks/load/ExportJob.java|0|1|00.00%|[369]|
|:large_blue_circle:|com/starrocks/sql/analyzer/ExpressionAnalyzer.java|1|22|04.55%|[843, 845, 846, 847, 848, 851, 852, 854, 855, 856, 857, 860, 861, 863, 864, 869, 872, 873, 876, 877, 881]|
|:large_blue_circle:|com/starrocks/qe/ConnectContext.java|1|5|20.00%|[279, 280, 281, 295]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/SqlToScalarOperatorTranslator.java|1|4|25.00%|[463, 484, 485]|
|:large_blue_circle:|com/starrocks/qe/SetExecutor.java|12|45|26.67%|[82, 85, 86, 133, 135, 136, 138, 139, 140, 142, 144, 145, 147, 148, 151, 152, 153, 154, 157, 158, 160, 163, 165, 166, 167, 168, 169, 170, 171, 172, 173, 174, 176]|
|:large_blue_circle:|com/starrocks/statistic/StatisticExecutor.java|7|20|35.00%|[72, 92, 96, 141, 143, 144, 145, 146, 147, 149, 237, 238, 239]|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|10|16|62.50%|[348, 1376, 1379, 1380, 1381, 1383]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/UserVariable.java|28|28|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/StreamLoadPlanner.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|12|12|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/DeletePlanner.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/load/loadv2/LoadingTaskPlanner.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/analysis/Expr.java|7|7|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/UpdatePlanner.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/StatementPlanner.java|13|13|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/PlanFragment.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/InsertPlanner.java|5|5|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/qe/ConnectContext.java [hunk 8]
```diff
@@ -118,6 +119,8 @@ public class ConnectContext {
     protected SessionVariable sessionVariable;
     // all the modified session variables, will forward to leader
     protected Map<String, SetVar> modifiedSessionVariables = new HashMap<>();
+    // user define variable in this session
+    protected HashMap<String, UserVariable> userVariables;
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/qe/ConnectContext.java:123]: will there be a lot of user variables?  do we support user use two variable at the same time?

**Comment by kangkaisen** [comment]
https://sonarcloud.io/project/issues?resolved=false&types=BUG&pullRequest=10011&id=StarRocks_starrocks&open=AYKhfMPJWq9q5u6cCyqK  The sonar issue should be fixed

**Comment by HangyuanLiu** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 113 / 191 (59.16%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/analysis/OutFileClause.java|0|1|00.00%|[227]|
|:large_blue_circle:|com/starrocks/load/ExportJob.java|0|1|00.00%|[369]|
|:large_blue_circle:|com/starrocks/sql/analyzer/ExpressionAnalyzer.java|1|22|04.55%|[843, 845, 846, 847, 848, 851, 852, 854, 855, 856, 857, 860, 861, 863, 864, 869, 872, 873, 876, 877, 881]|
|:large_blue_circle:|com/starrocks/qe/ConnectContext.java|1|7|14.29%|[280, 281, 282, 284, 285, 299]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/SqlToScalarOperatorTranslator.java|1|4|25.00%|[463, 484, 485]|
|:large_blue_circle:|com/starrocks/qe/SetExecutor.java|12|45|26.67%|[82, 85, 86, 133, 135, 136, 138, 139, 140, 142, 144, 145, 147, 148, 151, 152, 153, 154, 157, 158, 160, 163, 165, 166, 167, 168, 169, 170, 171, 172, 173, 174, 176]|
|:large_blue_circle:|com/starrocks/statistic/StatisticExecutor.java|9|20|45.00%|[92, 96, 141, 143, 144, 145, 146, 147, 149, 238, 239]|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|14|16|87.50%|[348, 1381]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/UserVariable.java|28|28|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/StreamLoadPlanner.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|12|12|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/DeletePlanner.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/load/loadv2/LoadingTaskPlanner.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/analysis/Expr.java|7|7|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/UpdatePlanner.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/StatementPlanner.java|13|13|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/PlanFragment.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/InsertPlanner.java|5|5|100.00%|[]|

**Review by Seaven — APPROVED** [review]

**Review by kangkaisen — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Pull request merged**